RA failure on seed=5474131906442374464

```
---- STDOUT -----------------------------------------
Making a frame with nrows=140, ncols=1
  types: bool=0, int=1, float=0, str=0
  missing values: 0.000
Launching an attack for 6 rounds
[02] Applying row slice (None, 128, 2) -> nrows = 64
[02] Applying row slice (6, None, 7) -> nrows = 9
[02] Applying row slice (1, None, 1) -> nrows = 8
[01] Setting nrows to 5
[02] Applying row slice (None, 0, -1) -> nrows = 4
[01] Setting nrows to 3
Attack end, checking the outcome

---- STDERR -----------------------------------------
Traceback (most recent call last):
  File "random_attack.py", line 400, in <module>
    ra.attack()
  File "random_attack.py", line 39, in attack
    frame.check()
  File "random_attack.py", line 286, in check
    self.df.internal.check()
AssertionError: Invalid min/max values in a Slice RowIndex 82/3/-14: min = 40, max = 82
```

MRE:
```
>>> import datatable as dt
>>> f0 = dt.Frame(range(25))
>>> f1 = f0[::2, :]
>>> f1.nrows = 3
>>> f1.internal.check()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Invalid min/max values in a Slice RowIndex 0/3/2: min = 0, max = 24
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -76,6 +76,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
   ids when stringified (#1241).
 - crash when saving a frame with many boolean columns into CSV (#1278).
 - incorrect .stypes/.ltypes property after calling cbind().
+- calculation of min/max values in internal rowindex upon row resizing.
 
 
 ### [v0.6.0](https://github.com/h2oai/datatable/compare/v0.6.0...v0.5.0) — 2018-06-05

--- a/c/rowindex.h
+++ b/c/rowindex.h
@@ -70,6 +70,8 @@ class ArrayRowIndexImpl : public RowIndexImpl {
   private:
     arr32_t ind32;
     arr64_t ind64;
+    bool is_sorted;
+    int64_t : 56;
 
   public:
     ArrayRowIndexImpl(arr32_t&& indices, bool sorted);
@@ -94,7 +96,7 @@ class ArrayRowIndexImpl : public RowIndexImpl {
     // Helper function that computes and sets proper `min` / `max` fields for
     // this RowIndex. The `sorted` flag is a hint whether the indices are
     // sorted (if they are, computing min/max is much simpler).
-    template <typename T> void set_min_max(const dt::array<T>&, bool sorted);
+    template <typename T> void set_min_max(const dt::array<T>&);
 
     // Helpers for `ArrayRowIndexImpl(Column*)`
     void init_from_boolean_column(BoolColumn* col);

--- a/c/rowindex_array.cc
+++ b/c/rowindex_array.cc
@@ -24,19 +24,21 @@
 ArrayRowIndexImpl::ArrayRowIndexImpl(arr32_t&& array, bool sorted)
     : ind32(std::move(array)), ind64()
 {
+  is_sorted = sorted;
   type = RowIndexType::RI_ARR32;
   length = static_cast<int64_t>(ind32.size());
   xassert(length <= std::numeric_limits<int32_t>::max());
-  set_min_max<int32_t>(ind32, sorted);
+  set_min_max<int32_t>(ind32);
 }
 
 
 ArrayRowIndexImpl::ArrayRowIndexImpl(arr64_t&& array, bool sorted)
     : ind32(), ind64(std::move(array))
 {
+  is_sorted = sorted;
   type = RowIndexType::RI_ARR64;
   length = static_cast<int64_t>(ind64.size());
-  set_min_max<int64_t>(ind64, sorted);
+  set_min_max<int64_t>(ind64);
 }
 
 
@@ -46,6 +48,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(
 {
   size_t n = starts.size();
   xassert(n == counts.size() && n == steps.size());
+  is_sorted = false;
 
   // Compute the total number of elements, and the largest index that needs
   // to be stored. Also check for potential overflows / invalid values.
@@ -102,6 +105,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(
 
 
 ArrayRowIndexImpl::ArrayRowIndexImpl(Column* col) {
+  is_sorted = false;
   switch (col->stype()) {
     case SType::BOOL:
       init_from_boolean_column(static_cast<BoolColumn*>(col));
@@ -120,6 +124,7 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(Column* col) {
 
 ArrayRowIndexImpl::ArrayRowIndexImpl(filterfn32* ff, int64_t n, bool sorted) {
   xassert(n <= std::numeric_limits<int32_t>::max());
+  is_sorted = sorted;
 
   // Output buffer, where we will write the indices of selected rows. This
   // buffer is preallocated to the length of the original dataset, and it will
@@ -199,11 +204,12 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(filterfn32* ff, int64_t n, bool sorted) {
   ind32.resize(out_length);
   length = static_cast<int64_t>(out_length);
   type = RowIndexType::RI_ARR32;
-  set_min_max(ind32, sorted);
+  set_min_max(ind32);
 }
 
 
 ArrayRowIndexImpl::ArrayRowIndexImpl(filterfn64* ff, int64_t n, bool sorted) {
+  is_sorted = sorted;
   size_t out_length = 0;
   int64_t rows_per_chunk = 65536;
   int64_t num_chunks = (n + rows_per_chunk - 1) / rows_per_chunk;
@@ -241,16 +247,16 @@ ArrayRowIndexImpl::ArrayRowIndexImpl(filterfn64* ff, int64_t n, bool sorted) {
   ind64.resize(out_length);
   length = static_cast<int64_t>(out_length);
   type = RowIndexType::RI_ARR64;
-  set_min_max(ind64, sorted);
+  set_min_max(ind64);
 }
 
 
 template <typename T>
-void ArrayRowIndexImpl::set_min_max(const dt::array<T>& arr, bool sorted) {
+void ArrayRowIndexImpl::set_min_max(const dt::array<T>& arr) {
   const T* data = arr.data();
   if (length <= 1) {
     min = max = (length == 0) ? 0 : static_cast<T>(data[0]);
-  } else if (sorted) {
+  } else if (is_sorted) {
     min = static_cast<T>(data[0]);
     max = static_cast<T>(data[length - 1]);
     if (min > max) std::swap(min, max);
@@ -289,7 +295,8 @@ void ArrayRowIndexImpl::init_from_boolean_column(BoolColumn* col) {
         if (data[i] == 1)
           ind32[k++] = static_cast<int32_t>(i);
       });
-    set_min_max(ind32, true);
+    is_sorted = true;
+    set_min_max(ind32);
   } else {
     type = RowIndexType::RI_ARR64;
     ind64.resize(zlen);
@@ -299,7 +306,8 @@ void ArrayRowIndexImpl::init_from_boolean_column(BoolColumn* col) {
         if (data[i] == 1)
           ind64[k++] = i;
       });
-    set_min_max(ind64, true);
+    is_sorted = true;
+    set_min_max(ind64);
   }
 }
 
@@ -359,7 +367,8 @@ RowIndexImpl* ArrayRowIndexImpl::uplift_from(RowIndexImpl* rii) {
         rowsres[i] = start + ind64[i] * step;
       }
     }
-    auto res = new ArrayRowIndexImpl(std::move(rowsres), false);
+    bool res_sorted = is_sorted && step >= 0;
+    auto res = new ArrayRowIndexImpl(std::move(rowsres), res_sorted);
     res->compactify();
     return res;
   }
@@ -371,7 +380,8 @@ RowIndexImpl* ArrayRowIndexImpl::uplift_from(RowIndexImpl* rii) {
     for (size_t i = 0; i < zlen; ++i) {
       rowsres[i] = rows_ab[rows_bc[i]];
     }
-    return new ArrayRowIndexImpl(std::move(rowsres), false);
+    bool res_sorted = is_sorted && arii->is_sorted;
+    return new ArrayRowIndexImpl(std::move(rowsres), res_sorted);
   }
   if (uptype == RowIndexType::RI_ARR32 || uptype == RowIndexType::RI_ARR64) {
     ArrayRowIndexImpl* arii = static_cast<ArrayRowIndexImpl*>(rii);
@@ -397,7 +407,8 @@ RowIndexImpl* ArrayRowIndexImpl::uplift_from(RowIndexImpl* rii) {
         rowsres[i] = rows_ab[rows_bc[i]];
       }
     }
-    auto res = new ArrayRowIndexImpl(std::move(rowsres), false);
+    bool res_sorted = is_sorted && arii->is_sorted;
+    auto res = new ArrayRowIndexImpl(std::move(rowsres), res_sorted);
     res->compactify();
     return res;
   }
@@ -472,12 +483,14 @@ RowIndexImpl* ArrayRowIndexImpl::inverse(int64_t nrows) const {
 
 void ArrayRowIndexImpl::shrink(int64_t n) {
   xassert(n < length);
+  length = n;
   if (type == RowIndexType::RI_ARR32) {
     ind32.resize(static_cast<size_t>(n));
+    set_min_max(ind32);
   } else {
     ind64.resize(static_cast<size_t>(n));
+    set_min_max(ind64);
   }
-  length = n;
 }
 
 RowIndexImpl* ArrayRowIndexImpl::shrunk(int64_t n) {
@@ -486,11 +499,11 @@ RowIndexImpl* ArrayRowIndexImpl::shrunk(int64_t n) {
   if (type == RowIndexType::RI_ARR32) {
     arr32_t new_ind32(zn);
     memcpy(new_ind32.data(), ind32.data(), zn * sizeof(int32_t));
-    return new ArrayRowIndexImpl(std::move(new_ind32), false);
+    return new ArrayRowIndexImpl(std::move(new_ind32), is_sorted);
   } else {
     arr64_t new_ind64(zn);
     memcpy(new_ind64.data(), ind64.data(), zn * sizeof(int64_t));
-    return new ArrayRowIndexImpl(std::move(new_ind64), false);
+    return new ArrayRowIndexImpl(std::move(new_ind64), is_sorted);
   }
 }
 
@@ -509,7 +522,7 @@ size_t ArrayRowIndexImpl::memory_footprint() const {
 
 template <typename T>
 static void verify_integrity_helper(
-    const dt::array<T>& ind, int64_t len, int64_t min, int64_t max
+    const dt::array<T>& ind, int64_t len, int64_t min, int64_t max, bool sorted
 ) {
   size_t zlen = static_cast<size_t>(len);
   if (ind.size() != zlen) {
@@ -518,6 +531,7 @@ static void verify_integrity_helper(
   }
   T tmin = std::numeric_limits<T>::max();
   T tmax = 0;
+  bool check_sorted = sorted;
   for (size_t i = 0; i < zlen; ++i) {
     T x = ind[i];
     if (ISNA<T>(x)) {
@@ -530,6 +544,7 @@ static void verify_integrity_helper(
     }
     if (x < tmin) tmin = x;
     if (x > tmax) tmax = x;
+    if (check_sorted && i > 0 && x < ind[i-1]) check_sorted = false;
   }
   if (!zlen) tmin = 0;
   if (tmin != min || tmax != max) {
@@ -538,15 +553,19 @@ static void verify_integrity_helper(
         << "/max=" << max << " compared to the computed min=" << tmin
         << "/max=" << tmax;
   }
+  if (check_sorted != sorted) {
+    throw AssertionError()
+        << "ArrrayRowIndex is marked as sorted, but actually it isn't.";
+  }
 }
 
 void ArrayRowIndexImpl::verify_integrity() const {
   RowIndexImpl::verify_integrity();
 
   if (type == RowIndexType::RI_ARR32) {
-    verify_integrity_helper<int32_t>(ind32, length, min, max);
+    verify_integrity_helper<int32_t>(ind32, length, min, max, is_sorted);
   } else if (type == RowIndexType::RI_ARR64) {
-    verify_integrity_helper<int64_t>(ind64, length, min, max);
+    verify_integrity_helper<int64_t>(ind64, length, min, max, is_sorted);
   } else {
     throw AssertionError() << "Invalid type = " << type << " in ArrayRowIndex";
   }

--- a/c/rowindex_slice.cc
+++ b/c/rowindex_slice.cc
@@ -198,6 +198,8 @@ RowIndexImpl* SliceRowIndexImpl::inverse(int64_t nrows) const {
 
 void SliceRowIndexImpl::shrink(int64_t n) {
   length = n;
+  if (step > 0) max = start + step * (n - 1);
+  if (step < 0) min = start + step * (n - 1);
 }
 
 RowIndexImpl* SliceRowIndexImpl::shrunk(int64_t n) {

--- a/tests/test_dt.py
+++ b/tests/test_dt.py
@@ -509,24 +509,29 @@ def test_resize_rows_api():
     f0 = dt.Frame([20])
     f0.nrows = 3
     f0.nrows = 5
+    f0.internal.check()
     assert f0.topython() == [[20, 20, 20, None, None]]
 
 
 def test_resize_rows0():
     f0 = dt.Frame(range(10), stype=dt.int32)
     f0.nrows = 6
+    f0.internal.check()
     assert f0.shape == (6, 1)
     assert f0.stypes == (dt.int32,)
     assert f0.topython() == [[0, 1, 2, 3, 4, 5]]
     f0.nrows = 12
+    f0.internal.check()
     assert f0.shape == (12, 1)
     assert f0.stypes == (dt.int32,)
     assert f0.topython() == [[0, 1, 2, 3, 4, 5] + [None] * 6]
     f0.nrows = 1
+    f0.internal.check()
     assert f0.shape == (1, 1)
     assert f0.stypes == (dt.int32,)
     assert f0.scalar() == 0
     f0.nrows = 20
+    f0.internal.check()
     assert f0.shape == (20, 1)
     assert f0.stypes == (dt.int32,)
     assert f0.topython() == [[0] * 20]
@@ -536,17 +541,21 @@ def test_resize_rows1():
     srcs = [[True], [5], [14.3], ["fooga"], ["zoom"]]
     stypes = (dt.bool8, dt.int64, dt.float64, dt.str32, dt.str64)
     f0 = dt.Frame(srcs, stypes=stypes)
+    f0.internal.check()
     assert f0.shape == (1, 5)
     assert f0.stypes == stypes
     f0.nrows = 7
+    f0.internal.check()
     assert f0.shape == (7, 5)
     assert f0.stypes == stypes
     assert f0.topython() == [src * 7 for src in srcs]
     f0.nrows = 20
+    f0.internal.check()
     assert f0.shape == (20, 5)
     assert f0.stypes == stypes
     assert f0.topython() == [src * 7 + [None] * 13 for src in srcs]
     f0.nrows = 0
+    f0.internal.check()
     assert f0.shape == (0, 5)
     assert f0.stypes == stypes
     assert f0.topython() == [[]] * 5
@@ -557,19 +566,23 @@ def test_resize_rows_nastrs():
     f0.nrows = 3
     assert f0.topython() == [["foo", None, None]]
     f0.nrows = 10
+    f0.internal.check()
     assert f0.topython() == [["foo"] + [None] * 9]
 
 
 def test_resize_view_slice():
     f0 = dt.Frame(range(100))
     f1 = f0[8::2, :]
+    f1.internal.check()
     assert f1.shape == (46, 1)
     assert f1.internal.isview
     f1.nrows = 10
+    f1.internal.check()
     assert f1.shape == (10, 1)
     assert f1.internal.isview
     assert f1.topython()[0] == list(range(8, 28, 2))
     f1.nrows = 15
+    f1.internal.check()
     assert f1.shape == (15, 1)
     assert not f1.internal.isview
     assert f1.topython()[0] == list(range(8, 28, 2)) + [None] * 5
@@ -578,14 +591,17 @@ def test_resize_view_slice():
 def test_resize_view_array():
     f0 = dt.Frame(range(100))
     f1 = f0[[1, 1, 2, 3, 5, 8, 13, 0], :]
+    f1.internal.check()
     assert f1.shape == (8, 1)
     assert f1.internal.isview
     assert f1.topython() == [[1, 1, 2, 3, 5, 8, 13, 0]]
     f1.nrows = 4
+    f1.internal.check()
     assert f1.shape == (4, 1)
     assert f1.internal.isview
     assert f1.topython() == [[1, 1, 2, 3]]
     f1.nrows = 5
+    f1.internal.check()
     assert f1.shape == (5, 1)
     assert not f1.internal.isview
     assert f1.topython() == [[1, 1, 2, 3, None]]

